src/card_operations/card_sign.c

```diff
@@ -0,0 +1,245 @@
+ * @param signature Pointer to the buffer where the signature will be stored.
+ * @return ISO7816 Returns SW_NO_ERROR if the signature generation was
+ * successful, or an error code if it failed.
+ */
+static ISO7816 get_card_auth_signature(uint8_t *sign_data,
+                                       uint16_t sign_data_size,
+                                       uint8_t *signature);
+
+/**
+ * @brief Initializes a smart card and signs data.
+ *
+ * @param card_data Pointer to the smart card operation data.
+ * @param sign_data Pointer to the data to be signed.
+ */
+static void handle_card_init_and_sign_data_operation(
+    card_operation_data_t *card_data,
+    card_sign_data_config_t *sign_data);
+
+/**
+ * @brief Handles the response of a smart card sign operation.
+ * Based on the error code, handles user action required and sets the corret UI
+ * screen.
+ *
+ * @param card_data Pointer to the smart card operation data.
+ * @param sign_data Pointer to the data to be signed.
+ * @return card_error_type_e Returns a card error code indicating the success or
+ * failure of the operation.
+ */
+static card_error_type_e handle_sign_data_operation_response(
+    card_operation_data_t *card_data,
+    card_sign_data_config_t *sign_data);
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static ISO7816 get_card_auth_signature(uint8_t *sign_data,
+                                       uint16_t sign_data_size,
+                                       uint8_t *signature) {
+  ASSERT(ECDSA_SIGNATURE_SIZE >= sign_data_size && NULL != sign_data &&
+         NULL != signature);
+
+  uint8_t sign_data_inout[ECDSA_SIGNATURE_SIZE] = {0};
+  uint16_t sign_data_size_inout = sign_data_size;
+  ISO7816 status = DEFAULT_UINT32_IN_FLASH_ENUM;
+  memcpy(sign_data_inout, sign_data, sign_data_size);
+  status = nfc_ecdsa(sign_data_inout, &sign_data_size_inout);
+
+  if (SW_NO_ERROR == status) {
+    memcpy(signature, sign_data_inout, ECDSA_SIGNATURE_SIZE);
+  }
+
+  return status;
+}
+
+static void handle_card_init_and_sign_data_operation(
+    card_operation_data_t *card_data,
+    card_sign_data_config_t *sign_data) {
+  card_data->nfc_data = init_nfc_connection_data(sign_data->family_id,
+                                                 sign_data->acceptable_cards);
+  card_initialize_applet(card_data);
+
+  if (CARD_OPERATION_SUCCESS == card_data->error_type) {
+    if (DEFAULT_UINT32_IN_FLASH == *(uint32_t *)(sign_data->family_id)) {
```

Comment:
`*(uint32_t *)` typecasting for the sake of reading is not good. alternatives: `U32_READ_BE_ARRAY`, `read_be`